File: VideoMergeAndroid/app/src/main/java/com/moon/videomerger/merger/FFmpegRunner.kt
package com.moon.videomerger.merger

import android.util.Log
import com.arthenica.ffmpegkit.FFmpegKit
import com.arthenica.ffmpegkit.FFmpegSession
import com.arthenica.ffmpegkit.ReturnCode
import com.arthenica.ffmpegkit.Statistics

/**
 * FFmpegKit 封装 —— 异步执行 ffmpeg 命令，提供进度和日志回调。
 */
class FFmpegRunner(
    private val command: String,
    private val totalDuration: Double,
    private val onProgress: (Float) -> Unit,
    private val onLog: (String) -> Unit,
    private val onComplete: (success: Boolean, message: String) -> Unit
) {
    companion object { private const val TAG = "FFmpegRunner" }

    fun execute() {
        FFmpegKit.executeAsync(command, { session ->
            handleComplete(session)
        }, { log ->
            val msg = log.message
            Log.d(TAG, msg)
            onLog(msg)
        }, { statistics ->
            handleStatistics(statistics)
        })
    }

    private fun handleComplete(session: FFmpegSession) {
        val rc = session.returnCode
        when {
            rc == null -> onComplete(false, "返回码为空")
            ReturnCode.isSuccess(rc) -> onComplete(true, "合并完成")
            ReturnCode.isCancel(rc) -> onComplete(false, "已取消")
            else -> {
                val logs = session.allLogsAsString ?: ""
                Log.e(TAG, "FFmpeg 失败: returnCode=${rc.value}\n$logs")
                onComplete(false, "合并失败 (code=${rc.value})")
            }
        }
    }

    private fun handleStatistics(statistics: Statistics) {
        if (totalDuration <= 0) return
        val currentSec = statistics.time / 1000.0
        val progress = (currentSec / totalDuration).toFloat().coerceIn(0f, 1f)
        onProgress(progress)
    }

    fun cancel() {
        FFmpegKit.cancel()
    }
}

File: VideoMergeIOS/VideoMerger/Merger/FFmpegRunner.swift
//
//  FFmpegRunner.swift
//  VideoMerger
//
//  FFmpegKit 异步封装 —— 提供进度与日志回调。
//  iOS 移植自 Android VideoMerger 的 FFmpegRunner.kt
//

import Foundation
import ffmpegkit

/// FFmpeg 执行结果
struct FFmpegResult {
    let success: Bool
    let message: String
    let returnCode: Int
}

/// FFmpeg 异步运行器
final class FFmpegRunner {

    private let command: String
    private let totalDuration: Double
    private let onProgress: (Float) -> Void
    private let onLog: (String) -> Void
    private let onComplete: (FFmpegResult) -> Void

    /// 当前会话（用于 cancel）
    private var session: FFmpegSession?

    init(command: String,
         totalDuration: Double,
         onProgress: @escaping (Float) -> Void,
         onLog: @escaping (String) -> Void,
         onComplete: @escaping (FFmpegResult) -> Void) {
        self.command = command
        self.totalDuration = totalDuration
        self.onProgress = onProgress
        self.onLog = onLog
        self.onComplete = onComplete
    }

    /// 启动异步执行
    func execute() {
        // 用 block 变量传递闭包，避免 Swift 闭包类型与 ObjC block 不匹配
        let completeBlock: FFmpegSessionCompleteCallback = { [weak self] session in
            self?.handleComplete(session)
        }
        let logBlock: LogCallback = { [weak self] log in
            guard let self = self, let log = log else { return }
            if let msg = log.getMessage() {
                self.onLog(msg + "\n")
            }
        }
        let statsBlock: StatisticsCallback = { [weak self] stats in
            self?.handleStatistics(stats)
        }

        session = FFmpegKit.executeAsync(
            command,
            withCompleteCallback: completeBlock,
            withLogCallback: logBlock,
            withStatisticsCallback: statsBlock
        )
    }

    /// 取消执行
    func cancel() {
        FFmpegKit.cancel()
    }

    // MARK: - 内部

    private func handleComplete(_ session: FFmpegSession?) {
        guard let session = session else {
            onComplete(FFmpegResult(success: false, message: "会话为空", returnCode: -1))
            return
        }
        guard let rc = session.getReturnCode() else {
            onComplete(FFmpegResult(success: false, message: "返回码为空", returnCode: -1))
            return
        }

        if ReturnCode.isSuccess(rc) {
            onComplete(FFmpegResult(success: true, message: "合并完成", returnCode: Int(rc.getValue())))
        } else if ReturnCode.isCancel(rc) {
            onComplete(FFmpegResult(success: false, message: "已取消", returnCode: Int(rc.getValue())))
        } else {
            let logs = session.getAllLogsAsString() ?? ""
            print("FFmpeg 失败: returnCode=\(rc.getValue())\n\(logs)")
            onComplete(FFmpegResult(
                success: false,
                message: "合并失败 (code=\(rc.getValue()))",
                returnCode: Int(rc.getValue())
            ))
        }
    }

    private func handleStatistics(_ stats: Statistics?) {
        guard let stats = stats, totalDuration > 0 else { return }
        let currentSec = stats.getTime()
        let progress = Float(max(0, min(1, currentSec / totalDuration)))
        onProgress(progress)
    }
}
